// @zag-js/pressable@0.6.0/dist/chunk-KJ5WME36.mjs
import {
  dom
} from "./chunk-3PRYODMI.mjs";
import {
  isHTMLAnchorLink,
  isOverTarget,
  isValidKeyboardEvent,
  shouldPreventDefaultKeyboard
} from "./chunk-QIHGLONP.mjs";

// src/pressable.machine.ts
import { createMachine, ref } from "@zag-js/core";
import { addDomEvent } from "@zag-js/dom-event";
import { isHTMLElement } from "@zag-js/dom-query";
import { disableTextSelection, restoreTextSelection } from "@zag-js/text-selection";
import { compact } from "@zag-js/utils";
function machine(userContext) {
  const ctx = compact(userContext);
  return createMachine(
    {
      id: "pressable",
      initial: "idle",
      context: {
        ...ctx,
        ignoreClickAfterPress: false,
        activePointerId: null,
        target: null,
        pointerType: null,
        pointerdownEvent: null,
        cleanups: ref([]),
        wasPressedDown: false
      },
      exit: ["restoreTextSelection", "removeDocumentListeners"],
      states: {
        idle: {
          entry: ["removeDocumentListeners", "resetContext", "restoreTextSelection", "resetIgnoreClick"],
          on: {
            POINTER_DOWN: [
              {
                guard: "isVirtualPointerEvent",
                actions: "setPointerType"
              },
              {
                target: "pressed:in",
                actions: [
                  "setPressedDown",
                  "setPointerType",
                  "setPointerId",
                  "setTarget",
                  "focusIfNeeded",
                  "disableTextSelection",
                  "invokeOnPressStart",
                  "trackDocumentPointerEvents"
                ]
              }
            ],
            KEY_DOWN: {
              target: "pressed:in",
              actions: ["setTarget", "invokeOnPressStart", "trackDocumentKeyup"]
            },
            CLICK: {
              actions: ["focusIfNeeded", "invokeOnPressStart", "invokeOnPressEnd", "invokeOnPress", "resetIgnoreClick"]
            }
          }
        },
        "pressed:in": {
          tags: "pressed",
          entry: "preventContextMenu",
          after: {
            500: {
              guard: "wasPressedDown",
              actions: ["clearPressedDown", "invokeOnLongPress"]
            }
          },
          on: {
            POINTER_LEAVE: [
              {
                guard: "cancelOnPointerExit",
                target: "idle",
                actions: ["clearPressedDown", "invokeOnPressEnd"]
              },
              {
                target: "pressed:out",
                actions: ["invokeOnPressEnd"]
              }
            ],
            DOC_POINTER_UP: [
              {
                guard: "wasPressedDown",
                target: "idle",
                actions: ["clearPressedDown", "invokeOnPressUp", "invokeOnPressEnd", "invokeOnPress"]
              },
              {
                target: "idle",
                actions: ["clearPressedDown", "invokeOnPressUp", "invokeOnPressEnd"]
              }
            ],
            DOC_KEY_UP: {
              target: "idle",
              actions: ["clearPressedDown", "invokeOnPressEnd", "triggerClick"]
            },
            KEY_UP: {
              target: "idle",
              actions: ["clearPressedDown", "invokeOnPressUp"]
            },
            DOC_POINTER_CANCEL: { target: "idle", actions: "clearPressedDown" },
            DRAG_START: { target: "idle", actions: "clearPressedDown" }
          }
        },
        "pressed:out": {
          on: {
            POINTER_ENTER: {
              target: "pressed:in",
              actions: "invokeOnPressStart"
            },
            DOC_POINTER_UP: {
              target: "idle"
            },
            DOC_POINTER_CANCEL: "idle",
            DRAG_START: "idle"
          }
        }
      }
    },
    {
      guards: {
        isVirtualPointerEvent: (_ctx, evt) => evt.pointerType === "virtual",
        cancelOnPointerExit: (ctx2) => !!ctx2.cancelOnPointerExit,
        wasPressedDown: (ctx2) => ctx2.wasPressedDown
      },
      actions: {
        trackDocumentPointerEvents(ctx2, _evt, { send }) {
          const doc = dom.getDoc(ctx2);
          const onPointerMove = (event) => {
            if (event.pointerId !== ctx2.activePointerId)
              return;
            const isOver = isOverTarget(event, ctx2.target);
            send({ type: isOver ? "POINTER_ENTER" : "POINTER_LEAVE", event });
          };
          const onPointerUp = (event) => {
            if (event.pointerId !== ctx2.activePointerId || event.button !== 0)
              return;
            send({ type: "DOC_POINTER_UP", event });
          };
          const onPointerCancel = (event) => {
            send({ type: "DOC_POINTER_CANCEL", event });
          };
          const cleanup = [
            addDomEvent(doc, "pointermove", onPointerMove, false),
            addDomEvent(doc, "pointerup", onPointerUp, false),
            addDomEvent(doc, "pointercancel", onPointerCancel, false)
          ];
          if (ctx2.pointerType !== "touch") {
            cleanup.push(addDomEvent(doc, "contextmenu", onPointerCancel, false));
          }
          ctx2.cleanups.push(...cleanup);
        },
        trackDocumentKeyup(ctx2, _evt, { send }) {
          const doc = dom.getDoc(ctx2);
          const onKeyup = (event) => {
            if (!isValidKeyboardEvent(event))
              return;
            if (shouldPreventDefaultKeyboard(event.target)) {
              event.preventDefault();
            }
            send({ type: "DOC_KEY_UP", event });
          };
          const cleanup = addDomEvent(doc, "keyup", onKeyup, false);
          ctx2.cleanups.push(cleanup);
        },
        removeDocumentListeners(ctx2) {
          ctx2.cleanups.forEach((fn) => fn?.());
          ctx2.cleanups = ref([]);
        },
        resetContext(ctx2) {
          ctx2.activePointerId = null;
          ctx2.pointerType = null;
          ctx2.pointerdownEvent = null;
        },
        restoreTextSelection(ctx2) {
          if (ctx2.allowTextSelectionOnPress || !ctx2.target)
            return;
          restoreTextSelection({ target: ctx2.target, doc: dom.getDoc(ctx2) });
        },
        setPointerToVirtual(ctx2) {
          ctx2.pointerType = "virtual";
        },
        setPointerType(ctx2, { event }) {
          ctx2.pointerType = event.pointerType;
        },
        setPointerId(ctx2, { event }) {
          ctx2.activePointerId = event.pointerId;
          ctx2.pointerdownEvent = ref(event);
        },
        setTarget(ctx2, { event }) {
          ctx2.target = ref(event.currentTarget);
        },
        focusIfNeeded(ctx2, { event }) {
          if (ctx2.disabled || ctx2.preventFocusOnPress)
            return;
          event.currentTarget.focus({ preventScroll: true });
        },
        invokeOnPressStart(ctx2, evt) {
          if (ctx2.disabled)
            return;
          let { event: originalEvent, pressEvent, pointerType } = evt;
          const event = pressEvent || originalEvent;
          ctx2.onPressStart?.({
            type: "pressstart",
            pointerType: pointerType || ctx2.pointerType,
            target: event.currentTarget,
            originalEvent: event
          });
        },
        disableTextSelection(ctx2) {
          if (!ctx2.target || ctx2.allowTextSelectionOnPress)
            return;
          disableTextSelection({ target: ctx2.target, doc: dom.getDoc(ctx2) });
        },
        invokeOnPressUp(ctx2, { event, pointerType }) {
          if (ctx2.disabled)
            return;
          ctx2.onPressUp?.({
            type: "pressup",
            pointerType: pointerType || ctx2.pointerType,
            target: event.currentTarget,
            originalEvent: event
          });
        },
        invokeOnPressEnd(ctx2, { event, pointerType }) {
          ctx2.ignoreClickAfterPress = true;
          ctx2.onPressEnd?.({
            type: "pressend",
            pointerType: pointerType || ctx2.pointerType,
            target: event.currentTarget,
            originalEvent: event
          });
        },
        invokeOnPress(ctx2, { event, pointerType }) {
          if (ctx2.disabled)
            return;
          ctx2.onPress?.({
            type: "press",
            pointerType: pointerType || ctx2.pointerType,
            target: event.currentTarget,
            originalEvent: event
          });
        },
        triggerClick(ctx2, { event }) {
          let target = event.target;
          if (!isHTMLElement(ctx2.target)) {
            return;
          }
          const isAnchor = isHTMLAnchorLink(ctx2.target) || ctx2.target.getAttribute("role") === "link";
          if (ctx2.target.contains(target) && isAnchor) {
            ctx2.target.click();
          }
        },
        dispatchPointerCancel(ctx2) {
          if (!ctx2.target)
            return;
          const win = dom.getWin(ctx2);
          const evt = new win.PointerEvent("pointercancel", { bubbles: true });
          ctx2.target.dispatchEvent(evt);
        },
        invokeOnLongPress(ctx2, { pointerType }) {
          if (!ctx2.target)
            return;
          ctx2.onLongPress?.({
            type: "longpress",
            pointerType: pointerType || ctx2.pointerType,
            target: ctx2.target,
            originalEvent: ctx2.pointerdownEvent
          });
        },
        resetIgnoreClick(ctx2) {
          ctx2.ignoreClickAfterPress = false;
        },
        setPressedDown(ctx2) {
          ctx2.wasPressedDown = true;
        },
        clearPressedDown(ctx2) {
          ctx2.wasPressedDown = false;
        },
        preventContextMenu(ctx2) {
          if (ctx2.pointerType !== "touch" || !ctx2.onLongPress)
            return;
          const onContextMenu = (event) => event.preventDefault();
          const cleanup = addDomEvent(ctx2.target, "contextmenu", onContextMenu, { once: true });
          const onPointerUp = () => void setTimeout(cleanup, 30);
          addDomEvent(dom.getWin(ctx2), "pointerup", onPointerUp, { once: true });
        }
      }
    }
  );
}

export {
  machine
};
