// solid-devtools@0.26.0/dist/index.js
import { startListeningWindowMessages, makeMessageListener, makePostMessage } from './chunk-6DJR4ASJ.js';
import { createInternalRoot, useDebugger, enableRootsAutoattach } from '@solid-devtools/debugger';
export { Debugger, attachDebugger, makeSolidUpdateListener, markComponentLoc, useDebugger, useLocator } from '@solid-devtools/debugger';
import { createEffect, onCleanup } from 'solid-js';

startListeningWindowMessages();
var _fromContent = makeMessageListener();
var fromContent = (...args) => onCleanup(_fromContent(...args));
var toContent = makePostMessage();
toContent("ResetPanel");
toContent("ClientConnected", "0.26.0");
var loadedBefore = false;
createInternalRoot(() => {
  const debug = useDebugger();
  fromContent("DevtoolsOpened", () => debug.toggleEnabled(true));
  fromContent("DevtoolsClosed", () => debug.toggleEnabled(false));
  createEffect(() => {
    if (!debug.enabled())
      return;
    if (loadedBefore)
      debug.emit("ResetState");
    else
      loadedBefore = true;
    fromContent((e) => debug.emit(e.name, e.details));
    debug.listen((e) => toContent(e.name, e.details));
  });
});
enableRootsAutoattach();
