// @zag-js/dom-query@0.1.4/dist/index.mjs
import {
  nextTick,
  raf
} from "./chunk-IWAFZYOG.mjs";
import {
  getComputedStyle
} from "./chunk-AK6CSOSE.mjs";
import {
  getEventTarget
} from "./chunk-V62Z6HGQ.mjs";
import {
  getParent,
  getScrollParent,
  getScrollParents
} from "./chunk-PHRRXCQY.mjs";
import {
  isEditableElement
} from "./chunk-67XKLLKN.mjs";
import {
  getPlatform,
  isApple,
  isDom,
  isFirefox,
  isIPhone,
  isIos,
  isMac,
  isSafari,
  isTouchDevice
} from "./chunk-ZPAS4QMN.mjs";
import {
  query,
  queryAll
} from "./chunk-5P7ELUHO.mjs";
import {
  ariaAttr,
  dataAttr
} from "./chunk-NX65SIAM.mjs";
import {
  contains
} from "./chunk-KP7WLHTU.mjs";
import {
  isHTMLElement
} from "./chunk-FKYCQ46C.mjs";
import {
  createScope
} from "./chunk-SE66MIC6.mjs";
import {
  getDocument,
  getWindow
} from "./chunk-5LKYLKZ7.mjs";
import {
  getActiveElement
} from "./chunk-MY2JRU5D.mjs";
import {
  getByTypeahead
} from "./chunk-HHMPWUPM.mjs";
import {
  getByText
} from "./chunk-CCVNLTYW.mjs";
import {
  indexOfId,
  itemById,
  nextById,
  prevById
} from "./chunk-HFFOD2VR.mjs";

// src/index.ts
var MAX_Z_INDEX = 2147483647;
export {
  MAX_Z_INDEX,
  ariaAttr,
  contains,
  createScope,
  dataAttr,
  getActiveElement,
  getByText,
  getByTypeahead,
  getComputedStyle,
  getDocument,
  getEventTarget,
  getParent,
  getPlatform,
  getScrollParent,
  getScrollParents,
  getWindow,
  indexOfId,
  isApple,
  isDom,
  isEditableElement,
  isFirefox,
  isHTMLElement,
  isIPhone,
  isIos,
  isMac,
  isSafari,
  isTouchDevice,
  itemById,
  nextById,
  nextTick,
  prevById,
  query,
  queryAll,
  raf
};
