// @zag-js/dom-event@0.6.0/dist/index.mjs
import {
  getRelativePointPercent,
  getRelativePointValue
} from "./chunk-L4YM5QC7.mjs";
import {
  requestPointerLock
} from "./chunk-HLSDWIYL.mjs";
import {
  trackPointerMove
} from "./chunk-UMOZC5FK.mjs";
import {
  trackVisualViewport
} from "./chunk-5TOYQE3U.mjs";
import {
  addDomEvent
} from "./chunk-UEVXZJ4S.mjs";
import {
  isContextMenuEvent,
  isCtrlKey,
  isKeyboardClick,
  isLeftClick,
  isModifiedEvent,
  isPrintableKey,
  isVirtualClick,
  isVirtualPointerEvent
} from "./chunk-DTOO4YSM.mjs";
import {
  fireBlurEvent,
  fireCustomEvent
} from "./chunk-4CMEY4E7.mjs";
import "./chunk-LYGZVQ73.mjs";
import {
  getEventKey
} from "./chunk-OY76JBTY.mjs";
import {
  getEventStep
} from "./chunk-HKJ5BW5U.mjs";
import {
  getNativeEvent
} from "./chunk-UGFZK6TG.mjs";
export {
  addDomEvent,
  fireBlurEvent,
  fireCustomEvent,
  getEventKey,
  getEventStep,
  getNativeEvent,
  getRelativePointPercent,
  getRelativePointValue,
  isContextMenuEvent,
  isCtrlKey,
  isKeyboardClick,
  isLeftClick,
  isModifiedEvent,
  isPrintableKey,
  isVirtualClick,
  isVirtualPointerEvent,
  requestPointerLock,
  trackPointerMove,
  trackVisualViewport
};
