// @solid-primitives/styles@0.0.101/dist/index.js
import { createSharedRoot } from '@solid-primitives/rootless';
import { createSignal, onCleanup } from 'solid-js';

// src/index.ts
var totallyHiddenStyles = {
  border: "0",
  padding: "0",
  visibility: "hidden",
  position: "absolute",
  top: "-9999px",
  left: "-9999px"
};
var getRemSize = () => parseFloat(getComputedStyle(document.documentElement).fontSize);
function createRemSize() {
  const [remSize, setRemSize] = createSignal(getRemSize());
  const el = document.createElement("div");
  Object.assign(el.style, totallyHiddenStyles, { width: "1rem" });
  document.body.appendChild(el);
  let init = true;
  const ro = new ResizeObserver(() => {
    if (init)
      return init = false;
    setRemSize(getRemSize());
  });
  ro.observe(el);
  onCleanup(() => {
    el.remove();
    ro.disconnect();
  });
  return remSize;
}
var useRemSize = /* @__PURE__ */ createSharedRoot(createRemSize);
var setServerRemSize = (size) => {
};

export { createRemSize, getRemSize, setServerRemSize, useRemSize };
