// @solid-primitives/styles@0.0.101/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rootless = require('@solid-primitives/rootless');
var solidJs = require('solid-js');

// src/index.ts
var totallyHiddenStyles = {
  border: "0",
  padding: "0",
  visibility: "hidden",
  position: "absolute",
  top: "-9999px",
  left: "-9999px"
};
var getRemSize = () => parseFloat(getComputedStyle(document.documentElement).fontSize);
function createRemSize() {
  const [remSize, setRemSize] = solidJs.createSignal(getRemSize());
  const el = document.createElement("div");
  Object.assign(el.style, totallyHiddenStyles, { width: "1rem" });
  document.body.appendChild(el);
  let init = true;
  const ro = new ResizeObserver(() => {
    if (init)
      return init = false;
    setRemSize(getRemSize());
  });
  ro.observe(el);
  solidJs.onCleanup(() => {
    el.remove();
    ro.disconnect();
  });
  return remSize;
}
var useRemSize = /* @__PURE__ */ rootless.createSharedRoot(createRemSize);
var setServerRemSize = (size) => {
};

exports.createRemSize = createRemSize;
exports.getRemSize = getRemSize;
exports.setServerRemSize = setServerRemSize;
exports.useRemSize = useRemSize;
